### ISSUE TITLE
Grpc Callback Losing Express Callstack/Context

### ISSUE THREAD

**Issue by mkeggen-iforium** [grpc/grpc-node#599]
### Problem description
Following a callback using the Grpc Client the callstack seems to lose the overall context and breaks out of the express pipeline.

Checking the callstack before the call and with a breakpoint within the callback highlights the issue.

```js
client.executeOurService(request, {deadline: timeout}, function(err, response) {
  // This callback has lost overall context 
});
```

### Environment
 - Linux
 - Node version 8.10.0
 - Node installation method npm
 - grpc version 1.15.1

### Additional context
Is there something we should be doing differently when creating the client or before we make the call to ensure that when the grpc client calls our callback that our context is retained?

**Comment by murgatroid99** [comment]
Can you clarify what you mean when you say that it has "lost overall context"? And what exactly is this express pipeline?

**Comment by nicolasnoble** [comment]
The callback itself is technically a closure, so any "context" it gets is technically captured within the closure. I'm as confused as @murgatroid99 as what you might mean there.

**Comment by mkeggen-iforium** [comment]
Essentially what seems to happen is prior to the GRPC client call the call stack shows the correct stack as I would expect.  After the GRPC call the callstack seems to have lost the main process, almost like a new process is handling the response.  I have attached screenshots below,

because of this behavior, middleware we have in our Node Express Pipeline is no longer trapping errors and other events which is causing other issues 

call stack - before GRPC call 
![image](https://user-images.githubusercontent.com/8982905/47666324-2e6b5500-db9b-11e8-9c8d-ebe35209eb8b.png)

call stack - after GRPC call
![image](https://user-images.githubusercontent.com/8982905/47666388-55c22200-db9b-11e8-8c53-be077c01523a.png)

**Comment by mkeggen-iforium** [comment]
I have just updated the first screen shot as it didn't quite show the scrollbars - so you can see the original callstack length is considerable compared to when the callback is fired

**Comment by nicolasnoble** [comment]
What you're describing is kind of typical javascript callback behavior. The callstack is interrupted because it's being triggered from within a native extension, but that's all. The only callstack you'd see above this would be the node event manager. Here's an example in pure javascript:

```js
function showError() {
  console.log(new Error().stack.split('\n'))
}

function main() {
  showError()
  setTimeout(showError, 1)
}

main()
```

Running this would output the following:

```
[ 'Error',
  '    at showError (/home/pixel/show-errors/callback.js:2:15)',
  '    at main (/home/pixel/show-errors/callback.js:6:3)',
  '    at Object.<anonymous> (/home/pixel/show-errors/callback.js:10:1)',
  '    at Module._compile (module.js:652:30)',
  '    at Object.Module._extensions..js (module.js:663:10)',
  '    at Module.load (module.js:565:32)',
  '    at tryModuleLoad (module.js:505:12)',
  '    at Function.Module._load (module.js:497:3)',
  '    at Function.Module.runMain (module.js:693:10)',
  '    at startup (bootstrap_node.js:191:16)' ]
[ 'Error',
  '    at Timeout.showError [as _onTimeout] (/home/pixel/show-errors/callback.js:2:15)',
  '    at ontimeout (timers.js:498:11)',
  '    at tryOnTimeout (timers.js:323:5)',
  '    at Timer.listOnTimeout (timers.js:290:5)' ]
```

As you can see, the second callstack comes from something that isn't "main". The root of the second callstack is node's event manager, and it's not showing up either.

Also, any sort of exception catching wouldn't work straight either. Example:

```js
function doException() {
  throw Error('foo')
}

function main() {
  try {
    doException()
  }
  catch(Error) {
    console.log('Caught exception')
  }
  try {
    setTimeout(doException, 1)
  }
  catch(Error) {
    console.log('Caught exception')
  }
}

main()
```

Yields the following:
```
Caught exception
/home/pixel/show-errors/exceptions.js:2
  throw Error('foo')
  ^

Error: foo
    at Timeout.doException [as _onTimeout] (/home/pixel/show-errors/exceptions.js:2:9)
    at ontimeout (timers.js:498:11)
    at tryOnTimeout (timers.js:323:5)
    at Timer.listOnTimeout (timers.js:290:5)
```

If you *can* trap asynchronous errors in your pipeline (such as something thrown using setTimeout as I am showing), then something else is wrong, but so far what you're describing to me is very typical javascript behavior.

**Comment by nicolasnoble** [comment]
Okay - could you build a small reproduction case that exhibits this?

**Comment by mkeggen-iforium** [comment]
Ok - further debugging I discovered a little more to the puzzle which actually led to a solution

We use Node Domain which was catching thrown errors.  When the Grpc callback occurs it seems to detach from that domain so our error is no longer caught.

I found this issue which had a workaround to wrap the callback in a `process.domain.bind`

Given node domain has been deprecated is this unlikely to be resolved within the grpc-node?

Related issue - https://github.com/grpc/grpc-node/issues/47
